### PR TITLE
CORE-1430: make the analysis sharing and unsharing endpoints asynchronous

### PR DESCRIPTION
**Pull request by slr71**

### CODE REVIEW

File: src/apps/service/apps.clj [hunk 1]
```diff
@@ -1,79 +1,21 @@
 (ns apps.service.apps
   (:use [apps.constants :only [de-system-id]]
-        [apps.service.oauth :only [authorization-uri has-access-token]]
+        [apps.service.apps-client :only [get-apps-client get-apps-client-for-username]]
```
Comment (slr71) [inline src/apps/service/apps.clj:3]: The code to create an apps client for the current user has been moved to a new namespace in order to make it easy to restart an asynchronous task if we ever decide we want to do that.

File: src/apps/service/apps.clj [hunk 2]
```diff
@@ -454,13 +396,21 @@
...
   (jobs/list-job-permissions (get-apps-client user) user job-ids params))
 
+(defn validate-job-sharing-request-body
+  [user sharing-requests]
+  (jobs/validate-job-sharing-request-body (get-apps-client user) user sharing-requests))
+
 (defn share-jobs
   [user sharing-requests]
-  {:sharing (jobs/share-jobs (get-apps-client user) user sharing-requests)})
+  (jobs/share-jobs (get-apps-client user) user sharing-requests))
```
Comment (slr71) [inline src/apps/service/apps.clj:405]: The data associated with an asynchronous task has to be a map, so the entire request body is now passed down.

File: src/apps/service/apps/jobs/sharing.clj [hunk 3]
```diff
@@ -150,42 +161,94 @@
   (process-job-inputs (partial share-input-file sharer sharee) apps-client job))
 
 (defn- share-job
-  [apps-client sharer sharee {job-id :analysis_id level :permission}]
-  (if-let [job (jp/get-job-by-id job-id)]
+  [update-fn apps-client sharer sharee {job-id :analysis_id level :permission}]
+  (let [job-id (uuidify job-id)
```
Comment (slr71) [inline src/apps/service/apps/jobs/sharing.clj:165]: The job ID here has been decoded from the retrieved asynchronous task, so it has to be converted to a UUID.

File: src/apps/service/apps/jobs/sharing.clj [hunk 4]
```diff
@@ -150,42 +161,94 @@
...
        (job-sharing-success job-id
                             job
                             level
                             (concat input-share-errs child-input-share-errs)
                             output-share-err-msg
                             app-share-err-msg))
      (catch [:type ::job-sharing-failure] {:keys [failure-reason]}
+       (update-fn (format "failed to share job ID %s with %s: %s" job-id sharee failure-reason))
        (job-sharing-failure job-id job level failure-reason))
-     (catch [:type ::permission-load-failure] {:keys [reason]}
```
Comment (slr71) [inline src/apps/service/apps/jobs/sharing.clj:175]: `::permission-load-failure` wasn't being thrown anywhere, so i removed this catch block.

File: src/apps/service/apps/jobs/sharing.clj [hunk 5]
```diff
@@ -150,42 +161,94 @@
...
+     (async-tasks/add-status async-task-id {:status "running"})
+     (let [apps-client (get-apps-client-for-username username)
+           user        (.getUser apps-client)]
+       (mapv (partial share-jobs-with-user (async-tasks/update-fn async-task-id "running") apps-client user)
+             (:sharing data))
+       (async-tasks/add-completed-status async-task-id {:status "completed"}))
+     (catch Object _
+       (log/error (:throwable &throw-context) "failed processing async task" async-task-id)
+       (async-tasks/add-completed-status async-task-id {:status "failed"})
+       (cn/send-general-analysis-sharing-failure-notification username async-task-id)))))
```
Comment (slr71) [inline src/apps/service/apps/jobs/sharing.clj:208]: There's a bug in the code if we ever get into this catch block.

File: src/apps/service/apps/jobs/sharing.clj [hunk 6]
```diff
@@ -150,42 +161,94 @@
...
+
 (defn share-jobs
-  [apps-client user sharing-requests]
-  (mapv (partial share-jobs-with-user apps-client user) sharing-requests))
+  [apps-client {username :shortUsername} sharing-requests]
+  (-> (log/spy :warn (async-tasks/new-task "analysis-sharing" username sharing-requests))
```
Comment (psarando) [inline src/apps/service/apps/jobs/sharing.clj:212]: Debug logging that can be removed now?
```suggestion
  (-> (async-tasks/new-task "analysis-sharing" username sharing-requests)
```

**Review by psarando — APPROVED** [review]
Very nice! ✨ 
LGTM 👍

File: src/apps/service/apps/jobs/sharing.clj (hunk 6 above)
Comment (slr71) [reply src/apps/service/apps/jobs/sharing.clj:212]: Ah, yep, I missed that. Thanks for catching it!

**Comment by slr71** [comment]
Thanks for the review!

**Pull request merged**